### PR TITLE
Extracts the PDF image replace logic into separate handlers

### PR DESCRIPTION
**Pull request by jmuscireum**
This allows for easy integration of custom logic (for example embedding blobs without the use of http).

Migration:
If internal resources (assets) are used via relative URLs these need to be prepended with resource://.

Barcodes and QR codes using the type attribute in the img tag are now instead generated by using the barcode:// and qr:// URI protocols.

### CODE REVIEW

File: src/main/java/sirius/web/templates/pdf/ImageReplacedElementFactory.java [hunk 1]
```diff
@@ -115,177 +82,30 @@ public ReplacedElement createReplacedElement(LayoutContext c,
...
-            // Length is uneven and no checksum will be added
-            if (length % 2 != 0 && !code.isGenerateChecksum()) {
-                return "0" + src;
-            }
-
-            // Length is even but a checksum will be added
-            if (length % 2 == 0 && code.isGenerateChecksum()) {
-                return "0" + src;
+    private PdfReplaceHandler findHandler(String protocol) {
+        for (PdfReplaceHandler handler : handlers) {
```
Comment (sabieber) [inline src/main/java/sirius/web/templates/pdf/ImageReplacedElementFactory.java:103]: Stream::findFirst -> orElseThrow?

File: src/main/java/sirius/web/templates/pdf/handlers/BarcodePdfReplaceHandler.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            Barcode code = new BarcodeInter25();
+            code.setGenerateChecksum(true);
+            return code;
+        }
+
+        throw new UnsupportedOperationException(Strings.apply("Type '%s' is not supported", type));
+    }
+
+    /**
+     * Pad the code if necessary.
```
Comment (sabieber) [inline src/main/java/sirius/web/templates/pdf/handlers/BarcodePdfReplaceHandler.java:109]: Pad -> Pads

File: src/main/java/sirius/web/templates/pdf/handlers/BarcodePdfReplaceHandler.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+import org.xhtmlrenderer.pdf.ITextFSImage;
+import sirius.kernel.commons.Strings;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.awt.Color;
+import java.awt.Image;
+
+/**
+ * Responsible for resolving barcode:// URIs.
```
Comment (sabieber) [inline src/main/java/sirius/web/templates/pdf/handlers/BarcodePdfReplaceHandler.java:26]: Resolves barcode:// URIs to ....

File: src/main/java/sirius/web/templates/pdf/handlers/HttpPdfReplaceHandler.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+package sirius.web.templates.pdf.handlers;
+
+import org.xhtmlrenderer.extend.FSImage;
+import org.xhtmlrenderer.extend.UserAgentCallback;
+import sirius.kernel.di.std.Register;
+
+import javax.annotation.Nullable;
+
+/**
+ * Responsible for resolving http:// and https:// URIs and resizing the image while maintaining the image ratio.
```
Comment (sabieber) [inline src/main/java/sirius/web/templates/pdf/handlers/HttpPdfReplaceHandler.java:18]: Resolves http:// and https:// URIs ....

File: src/main/java/sirius/web/templates/pdf/handlers/QrCodePdfReplaceHandler.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import org.xhtmlrenderer.extend.FSImage;
+import org.xhtmlrenderer.extend.UserAgentCallback;
+import org.xhtmlrenderer.pdf.ITextFSImage;
+import sirius.kernel.commons.Strings;
+
+import javax.annotation.Nullable;
+import java.awt.Color;
+
+/**
+ * Responsible for resolving qr:// URIs.
```
Comment (sabieber) [inline src/main/java/sirius/web/templates/pdf/handlers/QrCodePdfReplaceHandler.java:25]: Resolves qr:// URIs to ...

File: src/main/java/sirius/web/templates/pdf/handlers/ResourcePdfReplaceHandler.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+import sirius.kernel.commons.Strings;
+import sirius.kernel.di.std.Part;
+import sirius.kernel.di.std.Register;
+import sirius.web.resources.Resource;
+import sirius.web.resources.Resources;
+
+import javax.annotation.Nullable;
+
+/**
+ * Responsible for resolving resource:// URIs using {@link Resources} and resizing the image while maintaining the image
```
Comment (sabieber) [inline src/main/java/sirius/web/templates/pdf/handlers/ResourcePdfReplaceHandler.java:22]: Resolves ...

**Review by sabieber — CHANGES_REQUESTED** [review]

**Review by sabieber — APPROVED** [review]

**Review by idlira — APPROVED** [review]

**Review by mkeckmkeck — APPROVED** [review]

**Review by bwiedmann — APPROVED** [review]

**Review by andyHa — APPROVED** [review]

**Pull request merged**